### PR TITLE
52 bubble sort

### LINKED ISSUE #52: bubble_sort
**Description**
Implement a Bubble Sort algorithm to sort an array of integers in ascending order. Bubble Sort is a simple sorting algorithm that repeatedly steps through the list, compares adjacent elements, and swaps them if they are in the wrong order. The process is repeated until the list is sorted.

**Requirements:**
    The function should take a list of integers as input.
    The output should be a sorted list of integers in ascending order.

Example

     Input: [64, 34, 25, 12, 22, 11, 90]
    Output: [11, 12, 22, 25, 34, 64, 90]

**Additional Resources**

    [Bubble Sort Algorithm (Wikipedia)](https://en.wikipedia.org/wiki/Bubble_sort)
    [Sorting Algorithms Explained (YouTube)](https://www.youtube.com/)

### PR DESCRIPTION
**Pull request by Sufi-to**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/bubble_sort.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    assert all(
+        isinstance(el, (int, float)) for el in arr
+    ), "All elements in the list must be of type int or float."
+    assert not any(
+        isinstance(el, bool) for el in arr
+    ), "Boolean values are not allowed in the list."
+
+    for i in range(arrlen):
+        is_swaped = False
+        for j in range(arrlen - i - 1):
```
Comment (Nada-saad635) [inline solutions/bubble_sort.py:75]: here can you add comment explain this

**Comment by SiSaR-Pal** [comment]
The bubble_sort function is well-implemented, but I recommend adding explicit type annotations to enhance clarity and align with modern Python best practices. Specifically, update the function signature to include type annotations for the parameter and return type:

def bubble_sort(arr: list[float | int]) -> None:

File: solutions/bubble_sort.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+Module contents:
+    - bubble_sort(arr): Sorts an array in ascending order using the bubble sort algorithm.
+
+Created on 2025-1-4
+@author: Tomas Teclehaimanot
+"""
+
+
+def bubble_sort(arr):
```
Comment (SiSaR-Pal) [inline solutions/bubble_sort.py]: For Clarity:Add type annotations to specify that arr must be a list containing elements of type int or float, and that the function does not return anything (None).
```suggestion
def bubble_sort(arr: list[float | int]) -> None:
```

Comment (Sufi-to) [reply solutions/bubble_sort.py]: Can you check again I have already updated it.

Comment (SiSaR-Pal) [reply solutions/bubble_sort.py]: done

Comment (SiSaR-Pal) [reply solutions/bubble_sort.py]: Great work

**Review by SiSaR-Pal — APPROVED** [review]
Great Work

**Review by SiSaR-Pal — APPROVED** [review]
